arparser/src/parser.sk

```diff
@@ -0,0 +1,614 @@
+        this.reader.seek(IO.SeekFromStart(entry.data_start)) match {
+        | Failure(err) -> return Failure(IOError(err))
+        | Success _ -> void
+        };
+        this.reader.read_exact(buf.mbytes()) match {
+        | Failure(err) -> return Failure(IOError(err))
+        | Success _ -> void
+        };
+        return Success(Some(buf.bytes()))
+      }
+    };
+
+    Success(None())
+  }
+
+  mutable fun values(): mutable Iterator<Header> {
+    for (entry in this.headers) {
+      yield entry.header
+    }
+  }
+
+  mutable fun items(): mutable Iterator<(Header, readonly Bytes)> {
+    for (header in this) {
+      data = this.get(header.identifier) match {
+      | Success(Some(data)) -> data
+      | _ -> break void
+      };
+      yield (header, data)
+    }
+  }
+
+  mutable fun scan(): Result<void, Error> {
+    if (this.scanned) {
+      return Success(void)
+    };
+    this.parse_global_header() match {
+    | Failure(err) -> return Failure(err)
+    | Success _ -> void
+    };
+    loop {
+      header_start = this.next_entry_start;
+
+      this.reader.seek(IO.SeekFromStart(header_start)) match {
+      | Failure(err) -> return Failure(IOError(err))
+      | Success _ -> void
+      };
+
+      header = Header::read(this.reader) match {
+      | Failure(err) -> return Failure(err)
+      | Success(None()) -> break void
+      | Success(Some(v)) -> v
+      };
+
+      header_len = kENTRY_HEADER_LEN;
+      if (
+        !(this.variant is VariantBSD _) &&
+        header.identifier.startsWith("/")
+      ) {
+        this.!variant = VariantGNU();
+        if (header.identifier == kGNU_NAME_TABLE_ID) {
+          this.name_table.resize(header.size, UInt8::truncate(0));
+          this.reader.read_exact(this.name_table.mbytes()) match {
+          | Failure(err) -> return Failure(IOError(err))
+          | Success _ -> void
+          }
+        } else if (header.identifier != kGNU_SYMBOL_LOOKUP_TABLE_ID) {
+          !header.identifier = this.parse_gnu_extended_identifier(
+            header,
+          ) match {
+          | Failure(err) -> return Failure(err)
+          | Success(v) -> v
+          }
+        }
+      } else if (
+        !(this.variant is VariantBSD _) &&
+        header.identifier.endsWith("/")
+      ) {
+        this.!variant = VariantGNU();
+        !header.identifier = header.identifier.stripSuffix("/")
+      } else if (
+        !(this.variant is VariantGNU _) &&
+        header.identifier.startsWith(kBSD_EXTENDED_IDENTIFIER_PREFIX)
+      ) {
+        this.!variant = VariantBSD();
+        (
+          extended_identifier,
+          header_extra_len,
+        ) = this.parse_bsd_extended_identifier(header) match {
+        | Failure(err) -> return Failure(err)
+        | Success(v) -> v
+        };
+        !header.identifier = extended_identifier;
+        !header.size = header.size - header_extra_len;
+        !header_len = header_len + header_extra_len
+      };
+
```

Comment:
IIUC, we know from the very first file whether we're in the GNU (`//` or name ending with `/`) or BSD variant (anything else), right?

Reply:
Not really because BSD allows file names to contain `/`.

Reply:
Can a BSD filename **end** with `/`?

Reply:
According to the [FreeBSD man page for `ar (5)`](https://man.freebsd.org/cgi/man.cgi?query=ar&sektion=5), it does not seem to be forbidden. I'll re-check FreeBSD's `ar` implementation to be sure.

Reply:
Looking at [binutils](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=bfd/archive.c;h=9f3fbce9bc963fd72292451f85113cf33a8d002d;hb=HEAD#l591), they use the presence of a terminating `/` as an indication of `GNU/SysV` format. Whether BSD archives can _technically_ have a `/` at the end is an other question, but it seems a fringe enough use case that we should not worry about it.